jl_dlfind: do not find symbols in library dependencies

xal-0 opened the pull request:
A `ccall` without a specific library finds one with `jl_dlfind`, which calls
`jl_dlsym` on each of the following libraries, in order:
- `libjulia-internal`
- `libjulia`
- The current executable
- (On Windows): `kernel32`, `crtdll`, `ntdll`, `ws2_32`

The semantics of using `dlsym` (this does not apply to `GetProcAddress`) with a
specific handle are a little weird: the library and its dependencies are
searched, even if the provided symbol would resolve somewhere else if used in
the library.

On macOS and Linux, this causes all of the calls to libc functions in Base go
through the handle for `libjulia-internal`, making it difficult to hook
functions.  For example, `-fsanitize=thread` on clang intercepts malloc and free
by defining them in the executable; calls to malloc from Julia code would go to
the original libc version while calls to free in `libjulia-internal` would go to
the hooked version.

This change makes `jl_dlfind` return a handle only if the symbol is found in
that library and not one of its dependencies.

## Example

`a.c`:
```c
void b_func(void);

void c_func(void) {
  puts("c override from a");
}

int main() {
  puts("a");
  b_func();
}
```

`b.c`:
```c
define _GNU_SOURCE
#include <stdio.h>
#include <dlfcn.h>

void c_func(void);

void b_func(void) {
  puts("b");
  c_func();

  /* How jl_libjulia_internal_handle is set by jl_find_dynamic_library_by_addr */
  Dl_info info;
  dladdr(&b_func, &info);
  void *hdl = dlopen(info.dli_fname, RTLD_NOW | RTLD_NOLOAD | RTLD_LOCAL);
  void *(*dl_c_func)(void) = dlsym(hdl, "c_func");
  dl_c_func();
}
```

`c.c`:
```c
#include <stdio.h>

void c_func(void) {
  puts("c");
}
```

```
$ cc -g -fPIC -shared -o libc.so c.c
$ cc -g -fPIC -shared -o libb.so b.c libc.so
$ cc -Wl,-rpath,. -g -fPIC -o main a.c libb.so libc.so
$ ./main
a
b
c override from a
c
```

xal-0 committed:
jl_dlfind: do not find symbols in library dependencies

xal-0 committed:
Mark jl_find_dynamic_library_by_addr as JL_NOTSAFEPOINT

vtjnash reviewed (approved):
On macOS, you can additionally specify RTLD_FIRST when getting these internal handles and (like Windows) avoid the need for the extra dladdr call check

JeffBezanson commented:
I appreciate the very clear writeup given the unintuitive behavior here.

xal-0 committed:
Add missed no_deps jl_dlsym argument (Windows)

xal-0 committed:
Fix one other missing Windows jl_dlsym call

xal-0 commented:
> On macOS, you can additionally specify RTLD_FIRST when getting these internal handles and (like Windows) avoid the need for the extra dladdr call check

I thought it might be too breaking, but you're right that code relying on it wouldn't work on Windows anyhow. 

Looking a bit harder for options on Linux, it seems we can get the handle from an address directly with `dladdr1` with `RTLD_DL_LINKMAP` (apparently `dlopen` directly returns the `struct link_map *`as the handle?), but there are no nice options to make `dlsym` do the lookup we want.

topolarity commented:
> How do you do that? Does the implementation here need to be changed?

Yeah, it's required to mask off some of the lower bits (see https://github.com/opensource-apple/dyld/blob/3f928f32597888c5eac6003b9199d972d49857b5/src/dyldAPIs.cpp#L1513-L1518 and https://github.com/JuliaLang/julia/blob/4ae3f5ebed5378909f6355fb03b00057052c6766/src/sys.c#L668-L670)

If we use `RTLD_FIRST` though, the reverse lookup + handle comparison may no longer be necessary

ararslan commented on src/dlload.c:
```diff
@@ -403,7 +404,7 @@ JL_DLLEXPORT void *jl_load_dynamic_library(const char *modname, unsigned flags,
 }
 
-JL_DLLEXPORT int jl_dlsym(void *handle, const char *symbol, void ** value, int throw_err) JL_NOTSAFEPOINT
+JL_DLLEXPORT int jl_dlsym(void *handle, const char *symbol, void ** value, int throw_err, int no_deps) JL_NOTSAFEPOINT
```
It could be nice from a readability perspective to not have the argument represent a negation, as the double negative in `!no_deps` is (subjectively) less intuitive. That would also make it consistent with the `throw_err` argument, which isn't e.g. `dont_throw_err`. Anyway, just my drive-by $0.02, now back into the shadows I go...

xal-0 committed:
Make almost all jl_dlsym calls no_deps; use RTLD_FIRST on macOS

All calls to jl_dlsym from within Julia are now strict about which library the
symbol is from.  Now only Libdl.dlsym uses the default behaviour, to keep it the
same as libc dlsym.

On macOS, we can skip the jl_find_dynamic_library_by_addr check if we detect
that the handle was opened with RTLD_FIRST.  Do this by default for libraries
loaded by Julia ccalls.

xal-0 commented:
Reviving this! I've made both the internal `jl_dlsym` calls as well as ccalls use `no_deps=1`, which should result in less confusing differences between platforms. I also used @topolarity suggestion that we can avoid the `jl_find_dynamic_library_by_addr` check on macOS if we open our handles with `RTLD_FIRST` by default.

ararslan commented:
The FreeBSD failures look related:
```
Error in testset ccall:
Test Failed at /usr/home/julia/.buildkite-agent/builds/freebsd13-amdci6-2/julialang/julia-master/julia-8bd76994c8/share/julia/test/ccall.jl:1988
  Expression: malloc_hdl != ccall(:jl_dlfind, Ptr{Nothing}, (Cstring,), "jl_gc_safepoint")
   Evaluated: Ptr{Nothing}(0x0000000000000001) != Ptr{Nothing}(0x0000000000000001)
Error in testset ccall:
Test Failed at /usr/home/julia/.buildkite-agent/builds/freebsd13-amdci6-2/julialang/julia-master/julia-8bd76994c8/share/julia/test/ccall.jl:1989
  Expression: malloc_hdl != ccall(:jl_dlfind, Ptr{Nothing}, (Cstring,), "jl_array_ptr")
   Evaluated: Ptr{Nothing}(0x0000000000000001) != Ptr{Nothing}(0x0000000000000001)
```

xal-0 commented:
Investigating this has revealed all sorts of weird stuff:
- `jl_exe_handle` seems to be set to something sensible only on windows, where `GetModuleHandleA(NULL)` returns a handle to the current executable. On unix, we set `jl_exe_handle = jl_dlopen(NULL, JL_RTLD_NOW);`, which returns a handle that behaves like `RTLD_DEFAULT`.
-  However, the Windows-only`jl_RTLD_DEFAULT_handle = jl_libjulia_internal_handle;` makes no sense to me. As far as I can tell, we have no way of getting `RTLD_DEFAULT`-like behaviour on Windows with the current `jl_dlsym`. Stuff like `ccall(:memcmp, ...)` seems to have been working on unix because of the `RTLD_DEFAULT` fallback in `jl_dlfind`, while relying on the `msvcrt.dll` check on Windows.

xal-0 committed:
Rework RTLD_DEFAULT, jl_exe_handle

ararslan commented:
Different failure now on FreeBSD with the most recent commit:
```
Error in testset ccall:
Test Failed at /usr/home/julia/.buildkite-agent/builds/freebsd13-amdci6-3/julialang/julia-master/julia-11c20e9172/share/julia/test/ccall.jl:1991
  Expression: ccall(:jl_dlfind, Int, (Cstring,), "jl_gc_safepoint") == 2
   Evaluated: 0 == 2
```

xal-0 commented:
> Different failure now on FreeBSD with the most recent commit:

Turns out this was yet another linuxism we were relying on: you can't `dlclose` a handle but continue to use it for `dlsym` lookups on FreeBSD.

xal-0 committed:
Add close argument to jl_find_dynamic_library_by_addr

FreeBSD maintains two separate reference counts inside dlopen handles: one for
total references, including transitive dependencies loaded by rtld, and another
only for dlopen handles.  It checks to make sure that there are unclosed dlopen
handles when doing dlsym, so we've got to ensure that at least one handle
exists for all the libraries we want to use with
jl_find_dynamic_library_by_addr (specifically, libjulia and libjulia-internal).

xal-0 committed:
Merge branch 'master' into dlfind-strict

xal-0 committed:
Remove special error message for Windows RTLD_DEFAULT

ararslan commented:
Nice, FreeBSD is green now! However, Windows is not...
```
Error in testset ccall:
Test Failed at C:\buildkite-agent\builds\win2k22-amdci6-2\julialang\julia-master\julia-92411e7eef\share\julia\test\ccall.jl:1989
  Expression: ccall(:jl_dlfind, Int, (Cstring,), "main") == 1
   Evaluated: 0 == 1
```

xal-0 committed:
Fix broken jl_dlfind("main") test (should not run on Windows)

xal-0 committed:
Handle jl_find_dynamic_library_by_addr in main executable on BSD

FreeBSD returns NULL and no error when using dlopen() with RTLD_NOLOAD on the
main executable, so we can check and return jl_exe_handle directly.

xal-0 committed:
Rename and negate jl_dlsym parameter no_deps -> search_deps

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into dlfind-strict

vtjnash commented on src/dlload.c:
```diff
@@ -256,13 +257,19 @@ void *jl_find_dynamic_library_by_addr(void *symbol, int throw_err) {
         return NULL;
     }
     handle = dlopen(info.dli_fname, RTLD_NOW | RTLD_NOLOAD | RTLD_LOCAL);
```
```suggestion
    dlerror();
    handle = dlopen(info.dli_fname, RTLD_NOW | RTLD_NOLOAD | RTLD_LOCAL);
```

vtjnash commented on src/threading.c:
```diff
@@ -461,7 +461,7 @@ JL_DLLEXPORT jl_gcframe_t **jl_autoinit_and_adopt_thread(void)
     if (!jl_is_initialized()) {
         void *retaddr = __builtin_extract_return_addr(__builtin_return_address(0));
-        void *handle = jl_find_dynamic_library_by_addr(retaddr, 0);
+        void *handle = jl_find_dynamic_library_by_addr(retaddr, 0, 1);
```
```suggestion
        void *handle = jl_find_dynamic_library_by_addr(retaddr, 0, 0);
```

xal-0 committed:
Apply suggestions from code review

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into dlfind-strict

xal-0 committed:
On macOS, force RTLD_FIRST on dlsym handles instead of comparing

vtjnash committed:
Merge branch 'master' into dlfind-strict

vchuravy commented:
Just as a note, this has caused an observable behavior change in MPI.jl

https://github.com/JuliaParallel/MPI.jl/issues/915#issue-3418157479
